sdk/search/azure-search/src/main/java/com/azure/search/SearchIndexAsyncClient.java

```diff
@@ -0,0 +1,732 @@
+        HttpPipeline httpPipeline) {
+
+        SearchServiceUrlParts parts = SearchServiceUrlParser.parseServiceUrlParts(endpoint);
+
+        if (StringUtils.isBlank(indexName)) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid indexName"));
+        }
+        if (apiVersion == null) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid apiVersion"));
+        }
+        if (httpPipeline == null) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Invalid httpPipeline"));
+        }
+
+        this.endpoint = endpoint;
+        this.indexName = indexName;
+        this.apiVersion = apiVersion;
+        this.httpPipeline = httpPipeline;
+
+        restClient = new SearchIndexRestClientBuilder()
+            .searchServiceName(parts.serviceName)
+            .indexName(indexName)
+            .searchDnsSuffix(parts.dnsSuffix)
+            .apiVersion(apiVersion.getVersion())
+            .pipeline(httpPipeline)
+            .serializer(SERIALIZER)
+            .build();
+    }
+
+    /**
+     * Gets the name of the Azure Cognitive Search index.
+     *
+     * @return the indexName value.
+     */
+    public String getIndexName() {
+        return this.indexName;
+    }
+
+    /**
+     * Gets the {@link HttpPipeline} powering this client.
+     *
+     * @return the pipeline.
+     */
+    HttpPipeline getHttpPipeline() {
+        return this.httpPipeline;
+    }
+
+    /**
+     * Uploads a collection of documents to the target index
+     *
+     * @param documents collection of documents to upload to the target Index.
+     * @return document index result.
+     */
+    public Mono<IndexDocumentsResult> uploadDocuments(Iterable<?> documents) {
+        return this.uploadDocumentsWithResponse(documents)
+            .map(Response::getValue);
+    }
+
+    /**
+     * Uploads a collection of documents to the target index
+     *
+     * @param documents collection of documents to upload to the target Index.
+     * @return response containing the document index result.
+     */
+    public Mono<Response<IndexDocumentsResult>> uploadDocumentsWithResponse(Iterable<?> documents) {
+        return withContext(context -> uploadDocumentsWithResponse(documents, context));
+    }
+
+    @SuppressWarnings("unchecked")
+    Mono<Response<IndexDocumentsResult>> uploadDocumentsWithResponse(Iterable<?> documents, Context context) {
+        IndexBatch<?> batch = buildIndexBatch(documents, IndexActionType.UPLOAD);
+        return this.indexWithResponse(batch, context);
+    }
+
+    /**
+     * Merges a collection of documents with existing documents in the target index.
+     *
+     * @param documents collection of documents to be merged
+     * @return document index result
+     */
+    public Mono<IndexDocumentsResult> mergeDocuments(Iterable<?> documents) {
+        return this.mergeDocumentsWithResponse(documents)
```

Comment:
All public API methods that return Mono/Flux should have a try-catch block that catch all exceptions and pass it in the respective publisher's error channel. See [ConfigurationAsyncClient](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java#L124) for reference.

Reply:
added